### DISCUSSION TITLE
Image analysis

### DISCUSSION THREAD

**Discussion by SaharFarhangi** [neuropoly/axondeepseg#803] [Help needed]
![ADS](https://github.com/axondeepseg/axondeepseg/assets/163059145/2cbbe88a-b9b2-48f1-8ec9-a1a65ea1ad76)
I wanted to analyze the g-ratio but I faced this issue: FileNotFoundedError. I tried to upload and check the g-ratio of my friend's image, there was no issue.
![ADS](https://github.com/axondeepseg/axondeepseg/assets/163059145/8527a425-12af-4239-b978-c12964ff66f6)

**Comment by Stoyan-I-A** [comment]
Hello @SaharFarhangi,
In order to compute the morphometrics (such as the g-ratio), the software requires the segmentation masks of the axons and myelin of your image. Do you have these segmentation masks? If so, are they in the same place as your microscopy image and what are their files called?

**Comment by SaharFarhangi** [comment]
My problem is that it cannot recognize my image.

**Reply by Stoyan-I-A** [reply]
@SaharFarhangi From the screenshot that you showed, it seems that the software is looking for a file called `155A4_seg-axon.png`. The suffix `_seg-axon.png` is the one that AxonDeepSeg uses to save and identify the axon segmentation mask.
If no segmentation mask is present in the folder with the microscopy image, the morphometrics cannot be computed.

Could you describe the previous steps you performed before getting this error? Maybe this could help figuring out what went wrong.

**Comment by SaharFarhangi** [comment]
Sure. 
Turn on ADS plugin
Drag and drop grayscale inverted image into napari
Select model > rat TEM model
Settings > set zoom factor to 1 or 2 (I checked different zoom factor)
Apply ADS model > set pixel size to 0.069 micrometers per pixel

When I drag my colleague's image, it works without any errors.

**Reply by Stoyan-I-A** [reply]
@SaharFarhangi Thank you for the response. I have a few ideas on what could be the problem. I have a few questions:

1- What are the dimensions (in pixels) if your image?
The TEM model that we provide was trained with resolution of 0.01 micrometers per pixel. AxonDeepSeg would resample your image (which has a resolution of 0.069 micrometers per pixel)) before applying the model to try to have a resolution similar to what it was trained on, but in your case this might make the image too small and the software will fail to apply the model because the resampled image cannot fit a patch.

You could try to segment with a zoom factor of 0.14 and see if it helps.

2- If you get the error again, can you copy and paste the whole content of the terminal? This is useful information for the team to better understand where the error happens and if there are any indication of what could be the cause of the error.

**Comment by SaharFarhangi** [comment]
![ADS](https://github.com/axondeepseg/axondeepseg/assets/163059145/8151956c-18e7-4c13-964d-6f2928903cd2)
This is the dimension.

**Reply by Stoyan-I-A** [reply]
@SaharFarhangi Thank you for the information. As mentioned, you could try to segment using a zoom factor of  0.14

If you still get an error when trying to segment, please copy and paste the whole output of the terminal.

**Reply by mathieuboudreau** [reply]
Another option to debug would be to send us a copy of the image, and we could likely quickly find the optimal settings. The histology images look quite great, so I don't expect any issues getting the segmentation to work well and subsequently the morphometric values like g-ratio

**Comment by SaharFarhangi** [comment]
Thank you for your assistance. Zoom factor 0.14 is working. 
I have attached my image. Some axons are not fully filled with blue color (e.g., numbers 65 and 105). Should I fill them in? Additionally, I encountered some delamination, causing ADS to mistakenly identify them as separate axons (e.g., number 74). What do you suggest I do? 
![ADS new](https://github.com/axondeepseg/axondeepseg/assets/163059145/f00d8f9f-3c72-4b19-8b10-af59aecb4339)

**Reply by mathieuboudreau** [reply]
>Some axons are not fully filled with blue color (e.g., numbers 65 and 105). Should I fill them in? 

You can either to manual corrections (removing the myelin that was segmented inside the axon, and then filling it all in) or omit that axon in your analysis of the morphometrics (you can do so by looking at their ID numbers in the image above, and then not including those rows in the morphometrics file).

>Additionally, I encountered some delamination, causing ADS to mistakenly identify them as separate axons (e.g., number 74). What do you suggest I do?

Same thing, manual correction or omit; the choice is up to you, if you only have a few images to do then manual corrections is probably the best route, if you have many images then filtering them to be ignored in the morphometrics file is likely more time efficient

**Reply by SaharFarhangi** [reply]
Hi, I have encountered a recurring issue with my images, particularly with the larger axons, such as axon number 65. In this specific axon, there is a hole in the myelin and the myelin is not fully closed around the axon. I'm concerned about how to proceed without compromising my data.

If I delete this axon, I risk losing all the larger axons, which are crucial for my analysis. However, if I attempt to correct it by removing the segmented myelin inside the axon and filling it as an axon, I'm worried it could affect the myelin thickness measurements. Conversely, leaving it uncorrected might impact the g-ratio and other related metrics.

Could you please provide some guidance on how to handle this situation?

**Comment by SaharFarhangi** [comment]
Can AxonDeepSeg be used to visualize different myelin thicknesses with different colors, similar to the attached image? 
![ADS color](https://github.com/axondeepseg/axondeepseg/assets/163059145/ad84c99c-ec7a-41df-b10d-c09404c735e6)

**Reply by Stoyan-I-A** [reply]
Unfortunately, we do not have this feature in AxonDeepSeg at the moment.

**Reply by hermancollin** [reply]
well I think we have a PR for this (more or less) from @mathieuboudreau

**Reply by mathieuboudreau** [reply]
@hermancollin do you mean https://github.com/axondeepseg/axondeepseg/pull/781? I don't think my PR does that at the moment, but it could probably be integrated fairly easily

**Reply by hermancollin** [reply]
Yes that's what I meant. Weird, I thought we discussed "heatmaps" like in the original AxonSeg software but maybe we missed that.

**Comment by SaharFarhangi** [comment]
Given the 3 different magnifications I have for Axondeepseg analysis-4800x, 6800x, and 9300x- and considering that delamination is evident in my images, which magnification do you recommend based on your experience?

**Reply by hermancollin** [reply]
Hi @SaharFarhangi. If I understand correctly, the image you showed above was at 9300x, correct? If that's the case, it looks pretty good so you could continue with this magnification.

At the moment your images have a pixel size close to 0.1 um/px. In theory this is ideal for our TEM model.

**Comment by SaharFarhangi** [comment]
Hi, I have encountered a recurring issue with my images, particularly with the larger axons, such as axon number 65. In this specific axon, there is a hole in the myelin and the myelin is not fully closed around the axon. I'm concerned about how to proceed without compromising my data.

Based on @mathieuboudreau suggestion, If I delete this axon, I risk losing all the larger axons, which are crucial for my analysis. However, if I attempt to correct it by removing the segmented myelin inside the axon and filling it as an axon, I'm worried it could affect the myelin thickness measurements. Conversely, leaving it uncorrected might impact the g-ratio and other related metrics.
Does AxonDeepSeg detect myelin folds and correctly quantify the myelin sheets without including the area in the fold?

Could you please provide some guidance on how to handle this situation?
![ADS new](https://github.com/axondeepseg/axondeepseg/assets/163059145/03514d20-f202-4f0c-ab1a-abbad71ba91a)

**Reply by mathieuboudreau** [reply]
Hi @SaharFarhangi,

>Based on @mathieuboudreau suggestion, If I delete this axon, I risk losing all the larger axons,

No, you could selectively (by using the axon IDs) simply omit this one axon from your analysis

> However, if I attempt to correct it by removing the segmented myelin inside the axon and filling it as an axon, I'm worried it could affect the myelin thickness measurements. 

Hum, I don't think this is a worry - just erase the mask for that axon and manually segment it, it should then provide regular values for myelin thickness and size.

In general, your segmentation image looks quite good for most axons, this type of artifact can happen when there is some other dark features visible in the axon (i.e. if it doesn't look "regular" like what was in our training data).

The choice between using image as is but omitting IDs from the analysis or manually corrected the masks, to me it boils down to how many images do you have. If you only have a few (eg 10-30), then manually fixing these instances seems feasible time-wise. If however, you were to have hundreds of images, then I would first try omitting these from the statistics (reporting this in any publications), and you could select a few random images to see if there's a common feature in the ones you omitted (eg size, g-ratio) and do some aditional analysis to see if omitting these substantially impacted your analysis/results.

**Reply by SaharFarhangi** [reply]
Hi, thank you for your response.
I am analyzing the optic nerve, which include both small and large axons, across approximately 10 images for each sample. The issue with segmented myelin occurs specifically in the larger axons. I know I can delete the axons with the ID, but if I do this, I would have to delete all the large axons, which is a significant issue as they are crucial for my analysis.
Would you please explain more about it: just erase the mask for that axon and manually segment it, it should then provide regular values for myelin thickness and size.
Additionally, could you please take a look at axon number 58? It does not appear to be a separate axon but rather a segmented part related to axon number 59? How should I handle this situation?
For your review, I have attached 2 images at x6800 magnification, showing the segmentation before and after correction. I deleted the segmented myelin and filled it as an axon. Is it correct?
![4 after changing](https://github.com/axondeepseg/axondeepseg/assets/163059145/70abb782-b9b8-4495-8b4e-03566a80f065)
![4 before changing](https://github.com/axondeepseg/axondeepseg/assets/163059145/b5a02999-61b9-4cb0-b28a-991bb6aa4ffc)

**Reply by hermancollin** [reply]
@SaharFarhangi hello!

> Would you please explain more about it: just erase the mask for that axon and manually segment it, it should then provide regular values for myelin thickness and size.

If the model misses some axons, you can correct the mask before re-applying morphometrics. The manual correction you showed is a good example.

> Additionally, could you please take a look at axon number 58? It does not appear to be a separate axon but rather a segmented part related to axon number 59? How should I handle this situation?

Yes, unfortunately small glitches like this happen sometimes. You can correct the myelin mask + remove the spurious axon 58 in the axon mask.

> For your review, I have attached 2 images at x6800 magnification, showing the segmentation before and after correction. I deleted the segmented myelin and filled it as an axon. Is it correct?

Yes, overall this correction looks good. However I would be cautious when deleting the myelin to avoid under-segmentation.

**Comment by SaharFarhangi** [comment]
Hi, I am analyzing my TEM images using AxonDeepSeg. The magnification of my images is 4800x, and I have set the zoom factor to 0.14 and pixel size to 0.069. However, when I check the output for axon diameter and compare it with values from published papers, the axon diameter in my results appears to be 10 times larger than expected. Can you help me understand how I should adjust the settings in ADS to correctly account for magnification or scale bar? Thank you

**Reply by hermancollin** [reply]
Hi @SaharFarhangi Where does this value of 0.069 um/px come from? Is this the native resolution of the 4800x images?

**Comment by SaharFarhangi** [comment]
My colleague used this value for images at 100x magnification with toluidine blue staining. I applied the same value but adjusted the zoom factor from 4 to 0.14. I've attached one of my images at 4800x magnification for reference, but I converted it to JPG.
![A001](https://github.com/user-attachments/assets/e3801976-5e42-4912-a9ef-1387d9d2d24e)

**Reply by hermancollin** [reply]
@SaharFarhangi looking at the 2um scale bar, it seems the pixel size is 0.0196 um/px (also on the bottom left: 0.01943 um/px). This looks like the native resolution of the image.

If you use 0.0196 instead of 0.069, your values should be in the appropriate range. Can you confirm this?

**Reply by SaharFarhangi** [reply]
The bottom left is 0.01943. Should I use 0.01943 or 0.0196 um/px? Where does this value of 0.0169 um/px come from?

**Reply by mathieuboudreau** [reply]
>The bottom left is 0.01943. Should I use 0.01943 or 0.0196 um/px? Where does this value of 0.0169 um/px come from?

It comes from you, here it is in your comment yesterday: https://github.com/axondeepseg/axondeepseg/discussions/803#discussioncomment-11526581

**Reply by SaharFarhangi** [reply]
Previously, I used 0.069 µm/pixel, which was not accurate.
Does this mean that for each magnification, I should enter the specified µm/pixel value provided at the bottom of my images? For example, 0.01943 µm/pixel for 4800x magnification?

**Reply by mathieuboudreau** [reply]
Yes

**Reply by SaharFarhangi** [reply]
Thank you very much.

**Reply by SaharFarhangi** [reply]
Hi, how can I determine the appropriate zoom factor for an image captured at 4800x magnification (TEM) with 0.01943 um/pixel? I selected a zoom factor of 1, and it worked well, but I want to confirm if this is correct. Thank you.

**Reply by mathieuboudreau** [reply]
Hi @SaharFarhangi  - there is no exact way on determining this, it'll depend on the combination of the model being used (ie the data it was trained on) and the size of axons in your image. If a zoom factor of 1 gives you good results according to your eyes, then that's likely sufficient.

That being said, if you download the latest version of ADS, there is no longer a need for this zoom factor and it is no longer an option. So yo could also try that as an alternative.

**Reply by SaharFarhangi** [reply]
Hi, when the analysis is completed, each axon is labeled with a number, but there are some axons at the edges of the image where the number is not fully visible. For example, it shows 58, but the only digit 5 is clear. How can I determine the full number of such axons? Thank you.

**Reply by mathieuboudreau** [reply]
Hi @SaharFarhangi  - the numbers ascend in order (1,2,3,4...) going from left to right of the image, and then top to bottom, eg


1 2 3
4 5 6
7 8 9

If one of them is difficult to make out from deduction using this knowledge, you can look at the morphometrics file for the (x,y) pixel position of the center of each axon (top left of the image is (0,0))

**Reply by SaharFarhangi** [reply]
And how can I remove the axons located at the edge of the image? I want to keep the complete axons only. The ones at the edges of the image are incomplete.

**Reply by SaharFarhangi** [reply]
And how can I remove the axons located at the edge of the image? I want to keep the complete axons only. The ones at the edges of the image are incomplete.

**Reply by hermancollin** [reply]
@SaharFarhangi when you compute the morphometrics, you can use the `-b` option:

```axondeepseg_morphometrics -i <INPUT> -s <PIXELSIZE> -b```

which does the following:
```
  -b, --border-info     Adds a flag indicating if the axonmyelin object touches a border along with the 
                        coordinates of its bounding box.
```

A new column (called _border_touching_ I think) will indicate true if the axon touches the border of the image, and false otherwise. This way, you can conveniently filter these out of your analysis.

**Reply by SaharFarhangi** [reply]
Hi, when and where should I write this: 
axondeepseg_morphometrics -i <INPUT> -s <PIXELSIZE> -b
First, I upload my image, select TEM, apply ADS model and set pixel size, and finally compute morphmetrics.

**Reply by SaharFarhangi** [reply]
Sorry, I don't know how to use -b option.

**Reply by mathieuboudreau** [reply]
> Sorry, I don't know how to use -b option.

You type « -b » at the end of the axon morphometrics command.

**Reply by Stoyan-I-A** [reply]
The `-b` option is only available using the command line, not on napari.
So for example, you would write something like this in the command line:

`axondeepseg_morphometrics -i documents/path/to/your/image.tif -s 0.01 -b`

But we could add the border info to the napari plugin in the future as an option in the settings menu. 🤔

**Reply by SaharFarhangi** [reply]
I tried to write it in command line, but I couldn’t (seems inactive). The only things that I can write is 
First: Conda activate ads_venv
And then napari
Here you mean?

**Reply by SaharFarhangi** [reply]
here, I applied ADS model, and I want to start morphometrics, where should I add -b? I can't type anything.
![image](https://github.com/user-attachments/assets/d01e81e8-e66e-49c5-bace-21766d4bdd2a)

**Reply by Stoyan-I-A** [reply]
It would be before launching napari. In fact, it doesn't require napari at all.
First, use `conda activate ads_venv`
then, you can use the command `axondeepseg_morphometrics -i <INPUT> -s <PIXELSIZE> -b`
Note: replace `<INPUT>` by the path to your image and `<PIXELSIZE>` by the value of your pixel size.

**Comment by SaharFarhangi** [comment]
Hi. Thank you for your explanation. This is the path to my image:
![image](https://github.com/user-attachments/assets/57f9007b-31f0-4341-9f5e-3e3bb85db95c)
but it shows an error:
![image](https://github.com/user-attachments/assets/9012bad5-75d1-4c9c-97d0-8e0c9bf67c76)

**Reply by hermancollin** [reply]
Ah yes you should remove the `<>` around the file path and the pixel size, like so:
```
axondeepseg_morphometrics -i PATH -s 0.01943 -b
```

**Reply by SaharFarhangi** [reply]
I corrected based on your suggestion, but it shows error:
![image](https://github.com/user-attachments/assets/14843e3a-4a28-4270-95d9-469d8e0a37c6)

**Reply by mathieuboudreau** [reply]
It's trying to write the log file in the top disk folder (C:\), as that's where you're executing the function. This is typically protected on most operating systems, as you wouldn't want our software to be able to delete your entire operating system.

Try creating a working folder somewhere you have permissions, inside your User directory (or whatever Windows calls it), eg like your Documents folder

**Reply by mathieuboudreau** [reply]
Here, I quickly gave a few prompts to ChatGPT that wrote some additional step-by-step instructions to help you create a directory and move to that directory prior to running the command:

https://chatgpt.com/share/67895dca-c5ec-8009-a65c-c45ffef32e39

**Reply by SaharFarhangi** [reply]
I've made a new folder in the document, but shows this error for this step (Step 4: Use the New Directory in Your Script): 
![image](https://github.com/user-attachments/assets/8f289c88-fdfc-4364-a7ca-d2e49495d707)

**Reply by mathieuboudreau** [reply]
Don't bother with that step, now that you're in your proper directory, just do your initial call again from earlier, i.e. `axondeepseg_morphometrics -i Q:\Shared folders\Project NR1H3\EM\2024 Apr Sahar\Thin sections \Male \155A4\1
55A4 with KFCN 20241112 new and 20241219\ADS\4800x Crop\A001. tif -s 0.01943 -b`

**Comment by mathieuboudreau** [comment]
@SaharFarhangi it seems like you might need some more hands-on help to get this running for you - would you like to schedule a short zoom call so we could iron out in a single session? With screen sharing, I'm fairly sure we could resolve these mini-issues more efficiently

**Reply by SaharFarhangi** [reply]
Sure. Thank you. Are you available tomorrow after 12 PM Vancouver time (1/17/2025)?

**Reply by mathieuboudreau** [reply]
Send us an email to axondeepseg@googlegroups.com so that we can discuss scheduling and share a Zoom link

**Comment by SaharFarhangi** [comment]
Hi, I have some TEM images at 49000x with 0.001903 um/pix. when I applied ADS model at this pixel size. it didn't work. I saw this: the pixel size is too small for segmentation. How can I use ADS for these images? Thank you.

**Reply by mathieuboudreau** [reply]
Hi Sahar - if you are seeing that message, it means you’re using an old version of ADS. Please install the latest version, which doesnt require the pixel as an input to do the segmentation.

**Reply by mathieuboudreau** [reply]
As for the old software, its that the model is trying to resample the image to the resolution that we trained our TEM model on, and in doing so it’s making your image smaller than the size of our image patches. You could play around with the resolution value by increasing or decreasing it factors of 10 until the code works

**Comment by SaharFarhangi** [comment]
Hello, sorry yesterday I asked my question yesterday in different discussion chat which was not related to me.  We have successfully installed the new version of ADS. I have tested some images, but it is not recognizing all the axons. Do you have any recommendations to improve detection? I tried increasing contrast to see if it would help, but it did not seem to work.
Additionally, I want to confirm that when computing morphometrics, I should apply the pixel size corresponding to the magnification. Could you please clarify this?
Contrast is improved:
![image](https://github.com/user-attachments/assets/b6ae251f-a039-4f33-aa3d-c59657e82c18)
No contrast changing:
![image](https://github.com/user-attachments/assets/ebbac11a-4eb5-4d68-9d90-c223e48f1e61)
Another image: 
![image](https://github.com/user-attachments/assets/a5d041fd-d335-4c4c-b236-df88ed0186cb)

**Comment by mathieuboudreau** [comment]
Hi @SaharFarhangi,

Glad to hear you were able to get the software installed. 

Would you mind sharing the original image (without the segmentation overlayed) so that we can have a look?

**Reply by SaharFarhangi** [reply]
Hi @mathieuboudreau,
I've emailed the images. Thank you for your help.
